bpf/aya-bpf/src/maps/perf/perf_event_array.rs

```diff
@@ -50,20 +78,55 @@ impl<T> PerfEventArray<T> {
         }
     }
 
-    pub fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u32) {
-        self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)
+    pub fn output_current_cpu<C: BpfContext>(&self, ctx: &C, data: &T) -> Result<(), i64> {
+        self.output(ctx, data, BPF_F_CURRENT_CPU)
+    }
+
+    pub fn output_at_index<C: BpfContext>(&self, ctx: &C, data: &T, index: u32) -> Result<(), i64> {
+        self.output(ctx, data, u64::from(index) & BPF_F_INDEX_MASK)
     }
 
-    pub fn output_at_index<C: BpfContext>(&self, ctx: &C, index: u32, data: &T, flags: u32) {
-        let flags = u64::from(flags) << 32 | u64::from(index);
-        unsafe {
+    fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u64) -> Result<(), i64> {
```

Comment:
Having both `index` and `flags` looked subtly wrong, because the `flags` parameter of the bpf function is actually used either as an index, or as a flag, never both at the same time (and not with a 32-bits shift). If I'm not mistaken, either:
- you want to write at a specific `index: u32` and you set `flags = index as u64`
- or you want to write using a specific `flag: u64` and you set `flags = flag`

Reply:
> Having both index and flags looked subtly wrong, because the flags parameter of the bpf function is actually used either as an index, or as a flag

I don't think so? The flags argument can be used to pass _both_ an index and flags. The index just happens to be the low 32 bits. IOW, I might want to output to an index AND pass some flags

Reply:
I wonder if there are other flags than `BPF_F_CURRENT_CPU` for this call?
If I do `flags = u64::from(BPF_F_CURRENT_CPU) << 32 | u64::from(index)` the flags looks weird to me: which element is going to be written? In any case, `BPF_F_CURRENT_CPU` is not shifted in the bcc examples.

Reply:
> If I do flags = u64::from(BPF_F_CURRENT_CPU) << 32 | u64::from(index)
You get EINVAL, because you've set some invalid bits in the upper bits. 

>  In any case, BPF_F_CURRENT_CPU is not shifted in the bcc examples

Neither do we

> I wonder if there are other flags than BPF_F_CURRENT_CPU for this call?
It's a generic flags argument. More flags could be added in the future, and if that were to happen it shouldn't break our APIs

Reply:
> Neither do we

Didn't `self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)` discard the higher 32-bits of `BPF_F_CURRENT_CPU` and then apply the shift?

I understand the need for a generic future-proof `flag`, I will add it back.

Reply:
> > Neither do we
> 
> Didn't `self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)` discard the higher 32-bits of `BPF_F_CURRENT_CPU` and then apply the shift?

```
        let flags = u64::from(flags) << 32 | u64::from(index);
```

Yes but the shift is applied to flags, not index (BPF_F_CURRENT_CPU).

Reply:
it is if you look at _MASK and consider that kernel APIs are stable so the mask can't change

Reply:
\*revelation moment\* Okay, I understand now, sorry for the confusion. The API confused me a little bit.
May I offer to add the flags back, but with
```rust
let flags_for_bpf = (u64::from(index) & BPF_F_INDEX_MASK) | flags;
```
so that it's more in line with the documentation and will continue to work even if new flags are added?